crates/containerd-shim-wasmer/src/executor.rs

```diff
@@ -0,0 +1,132 @@
+use containerd_shim_wasm::sandbox::oci::{self, Spec};
+use containerd_shim_wasm::sandbox::Stdio;
+use libcontainer::workload::{Executor, ExecutorError, ExecutorValidationError};
+use wasmer::{Cranelift, Module, Store};
+use wasmer_wasix::WasiEnv;
+
+use crate::oci_wasmer;
+
+const EXECUTOR_NAME: &str = "wasmer";
+
+#[derive(Clone)]
+pub struct WasmerExecutor {
+    stdio: Stdio,
+    engine: Cranelift,
+}
+
+impl WasmerExecutor {
+    pub fn new(stdio: Stdio, engine: Cranelift) -> Self {
+        Self { stdio, engine }
+    }
+}
+
+impl Executor for WasmerExecutor {
+    fn exec(&self, spec: &Spec) -> Result<(), ExecutorError> {
+        match can_handle(spec) {
+            Ok(()) => {
+                let args = oci::get_args(spec);
+                if args.is_empty() {
+                    return Err(ExecutorError::InvalidArg);
+                }
+
+                let status = self
+                    .start(spec, args)
+                    .map_err(|err| ExecutorError::Other(format!("failed to prepare: {}", err)))?;
+
+                std::process::exit(status)
+            }
+            Err(ExecutorValidationError::CantHandle(_)) => {
+                LinuxContainerExecutor::new(self.stdio.clone()).exec(spec)?;
+
+                Ok(())
+            }
+            Err(_) => Err(ExecutorError::InvalidArg),
+        }
+    }
+
+    fn validate(&self, spec: &Spec) -> std::result::Result<(), ExecutorValidationError> {
+        match can_handle(spec) {
+            Ok(()) => Ok(()),
+            Err(ExecutorValidationError::CantHandle(_)) => {
+                LinuxContainerExecutor::new(self.stdio.clone()).validate(spec)?;
+
+                Ok(())
+            }
+            Err(err) => Err(err),
+        }
+    }
+}
+
+impl WasmerExecutor {
+    fn start(&self, spec: &Spec, args: &[String]) -> anyhow::Result<i32> {
+        log::info!("get envs from spec");
+        let envs = oci_wasmer::env_to_wasi(spec);
+
+        log::info!("redirect stdio");
+        self.stdio.take().redirect()?;
+
+        log::info!("get module_name and method");
+        let (module_name, _method) = oci::get_module(spec);
+        let module_name = match module_name {
+            Some(m) => m,
+            None => {
+                return Err(anyhow::format_err!(
+                    "no module provided, cannot load module from file within container"
+                ))
+            }
+        };
+
+        log::info!("Create a Store");
+        let mut store = Store::new(self.engine.clone());
+
+        log::info!("loading module from file {} ", module_name);
+        let module = Module::from_file(&store, module_name)?;
+
+        log::info!("Run the module args: {:?}, envs: {:?}", args, envs);
+        let result = WasiEnv::builder(EXECUTOR_NAME)
+            .args(args[1..].to_vec())
+            .envs(envs)
+            .preopen_dir("/")?
+            .run_with_store(module, &mut store);
```

Comment:
**todo:** `run_with_store` will only run `_start`. There's an example [here](https://github.com/wasmerio/wasmer/blob/94f27f81f601773071bd64dcd91c4c6b0d7d4196/examples/wasi_manual_setup.rs) showing how to run other entry functions:



```rust
        log::info!("loading module from file {module_name:?} ");
        let module = Module::from_file(&store, &module_name)?;

        let runtime = tokio::runtime::Builder::new_multi_thread()
            .enable_all()
            .build()?;
        let _guard = runtime.enter();

        log::info!("Run the module args: {:?}, envs: {:?}", args, envs);
        let (instance, env) = WasiEnv::builder(EXECUTOR_NAME)
            .args(&args[1..])
            .envs(envs)
            .preopen_dir("/")?
            .instantiate(module, &mut store)?;

        log::info!("Running {module_name:?}, {method:?}");
        let start = instance.exports.get_function(&method)?;
        env.data(&store).thread.set_status_running();
        let status = start.call(&mut store, &[]).map(|_| 0).or_else(|err| {
            match err.downcast_ref::<WasiError>() {
                Some(WasiError::Exit(code)) => Ok(code.raw()),
                _ => Err(err),
            }
        })?;

        Ok(status)
```

Reply:
I am happy for this to go in a different PR

Reply:
Thanks for your advice. I'll fix these after the merge.
I'll stay tuned.

Reply:
You're right, because I saw this:
https://github.com/wasmerio/wasmer/blob/3f5ff852ddf283b0921722e572c28739240be89e/lib/wasix/src/state/builder.rs#L819-L822